develop/tutorials/articles-dxp/310-troubleshooting/13-patching-and-debugging-portal-source-code.markdown

```diff
@@ -0,0 +1,192 @@
+
+From the
+[customer portal](https://web.liferay.com/group/customer/dxp),
+download a DXP Digital Enterprise (DXP) bundle (or DXP JARs), and the DXP source
+code for the DXP version you're using. For example, here are ZIP files for a DXP /
+Tomcat bundle and DXP code:
+
+-   `liferay-dxp-digital-enterprise-tomcat-7.0-sp4-20170705142422877.zip`
+-   `liferay-dxp-digital-enterprise-src-7.0-sp4-20170814143552654.zip`
+
+Download fix packs and their source code from
+[here](https://web.liferay.com/group/customer/dxp/downloads/digital-enterprise/fix-packs).
+For example, here are fix pack and source code  ZIP files:
+
+-   `liferay-fix-pack-de-29-7010.zip`
+-   `liferay-fix-pack-de-29-7010-src.zip`
+
+Next, you'll install and configure DXP first since it has the patching tool. If
+you have a patched DXP installation already and want to debug it, skip to the
+section on
+[setting up and patching the DXP source code](#prepare-the-dxp-source-code).
+
+## Step 2: Prepare DXP [](id=prepare-dxp)
+
+Setting up and patching DXP locally is straightforward. 
+
+### Install and Configure DXP [](id=install-and-configure-dxp)
+
+Follow these steps to install and configure DXP: 
+
+1.   [Install and Deploy DXP](/discover/deployment/-/knowledge_base/7-0/deploying-liferay-portal)
+locally. 
+
+2.  Start DXP.
+
+3.  Configure DXP to use your database. 
+
+4.  Stop DXP.
+
+It's time apply patches you want to DXP.
+
+### Patch DXP [](id=patch-dxp)
+
+Here are the steps for patching DXP:
+
+1.  Copy the patch ZIP files to the folder
+    `[LIFERAY_HOME]/patching-tool/patches`.
+
+2.  Open a terminal to the folder `[LIFERAY_HOME]/patching-tool`.
+
+3.  Generate a default patching profile called `default.properties` by running
+    `patching-tool.sh auto-discovery`. Make sure the profile's properties
+    refers to your DXP installation. See
+    [the patching tool documentation](/discover/deployment/-/knowledge_base/7-0/patching-tool)
+    for more details.
+
+    Here's an example profile:
+
+    	patching.mode=binary
+    	war.path=../tomcat-8.0.32/webapps/ROOT/
+    	global.lib.path=../tomcat-8.0.32/lib/ext/
+    	liferay.home=../
+
+4.  Execute the command `patching-tool.sh info` to list your installed
+    patches and patches available in your `patches` folder.
+
+5.  Install all available local patches by executing this command:
+
+        patching-tool.sh install
+
+    The
+    [patching tool documentation](/discover/deployment/-/knowledge_base/7-0/patching-tool)
+    explains additional steps that might apply, such as deleting your
+    `[LIFERAY_HOME]/osgi/state` folder and creating database indexes.
+
+You're ready to prepare the DXP source code for debugging. 
+
+## Step 3: Prepare the DXP Source Code [](id=prepare-the-dxp-source-code)
+
+Version control systems, such as [Git](https://git-scm.com/) facilitate
+examining code changes. To help you compare patches, it's recommended to commit
```

Comment:
This should be a secondary option. Patching tool has a feature which can diff patches and provide the information.